### ISSUE TITLE
Stepwise Planner (C#) does not execute action (treats and logs it as a thought)

### ISSUE THREAD

**Issue by gingters** [microsoft/semantic-kernel#2067]
**Describe the bug**

When the model executes a specific action on a plugin in the Stepwise planner, the planner does not recognize this as a function call, but logs the Action response from the model as a thought instead.

Maybe it has something to do with the fact that this specific plugin methods takes a json string as input?

Further context:
I wrote some plugins that allow control of a demo Blazor WASM application. The model can navigate the pages. A page that is currently visible can register a model for a form that is currently displayed with the plugin,  so that the model can ask the plugin for a list of forms. All these actions have no or only primitive parameters.
Then the model can see and fill out form fields on that model instance. This action takes fields with values for multiple forms, so the parameter type would be a `Dictionary<string, Dictionary<string, string>>`. This should be passed as a JSON string and I parse it in the dictionary in my plugin method.

I noticed following behavior on every single occasion I tested, so it seems to be very reproducible:

I ask the model that I want to create a new user with a given email and password.
Action: The model asks the plugin what screens the application provides.
It observes the list of screens and picks the correct screen to create a user.
Action: The model tells the plugin to navigate to the screen.
It observes that the application navigated to that screen and thinks it could check if there is a form to fill out.
Action: The model tells the plugin to list the available forms.
It observes the list of forms, and finds that the current form indeed has fields that it could fill out.
Action: The model tells the plugin to fill out the given form fields.

_Error_: Semantic Kernel interprets this action as a thought and does not execute the action. (See log further down).

**To Reproduce**

The demo application is not ready to be published, but I can give you individual access to the repo to reproduce the issue on the actual application.

But maybe the log is enough to spot the issue.

**Logs**
```
info: Joshua.Client.Features.AI_Integration.JoshuaAI[0]
      Plan created: Microsoft.SemanticKernel.SkillDefinition.FunctionView

      Ask: 
              You are Joshua, an AI assistant agent embedded in an application that is mainly used for Customer Relationship Management, but it also provides some general functionalities that do not have to do with CRM.
              You have partial control over some functionality of the application, provided to you through functions.
              You are able to navigate the application and select screens for the user, and sometimes it may be possible to extract information from the users request and use it to fill in a form after navigating there.
              You probably want to start by listing the avilable screens and their functionalities to check if there is a page available that can help the user to do what he wants.
              Give your final answer in the same language the user used to ask the question.
      
              User: Ich möchte einen Benutzer namens Anton mit der E-Mail Adresse anton@test.de und dem Passwort wrdlbrmpft anlegen.
      
      Thought: [THOUGHT PROCESS]
      [QUESTION]
      Ich möchte einen Benutzer namens Anton mit der E-Mail Adresse anton@test.de und dem Passwort wrdlbrmpft anlegen.

      Action: No action to take

      Thought: The user wants to create a new user named Anton with the email address anton@test.de and the password wrdlbrmpft. I should first list the available screens to see if there is a screen that can help with this task.

      Action: ApplicationNavigationPlugin.GetScreens({"INPUT":""})

      Observation: Here is a list of the screens that are available in the application. You can navigate to any of them by calling the 'NavigateTo' function with the route of the screen you want to navigate to. Please make sure to really navigate to a page before telling the user you did so, if you found a suitable screen.[{"ScreenName":"Counter","ScreenRoutes":["/counter/","/counter/{currentCount:int}"],"ScreenDescription":"This is a counter page. You can use it to count things. The initial counter value can be set by specifying the currentCount parameter in the URL."},{"ScreenName":"FetchData","ScreenRoutes":["/fetchdata"],"ScreenDescription":"This screen shows weather forecast data from the server."},{"ScreenName":"Index","ScreenRoutes":["/"],"ScreenDescription":"This is the main entry page into the application."},{"ScreenName":"UndoRedo","ScreenRoutes":["/undoredo"],"ScreenDescription":"A page with a counter. This demonstrates the use of undo/redo actions."},{"ScreenName":"PersonOverview","ScreenRoutes":["/personOverview"],"ScreenDescription":"Displays a list of persons with their skills and duties."},{"ScreenName":"PersonDetails","ScreenRoutes":["/personDetails/{EntryId:int}"],"ScreenDescription":"This page shows the details of a person based on the id of the person."},{"ScreenName":"Form","ScreenRoutes":["/form"],"ScreenDescription":"This is a form that shows how to use the MudBlazor MudForm component without a form model."},{"ScreenName":"Form2","ScreenRoutes":["/form2"],"ScreenDescription":"This is a form that lets you order two different things that are not further specified."},{"ScreenName":"ModelForm","ScreenRoutes":["/form3"],"ScreenDescription":"This is a form that lets you register a new user."},{"ScreenName":"UndoRedoForm","ScreenRoutes":["/form4"],"ScreenDescription":"This is a demonstration form that shows how to use the undo/redo functionality with a MudBlazor form. Only use this form if the user explicitly asks for undo/redo functionality."},{"ScreenName":"SimplerUndoRedoForm","ScreenRoutes":["/form5"],"ScreenDescription":"This is a form where you can change user profile information, including first name, last name, email and the password."}]

      Thought: There is a screen called "ModelForm" with the route "/form3" that allows registering a new user. I should navigate to that screen and then fill in the form with the provided information.

      Action: ApplicationNavigationPlugin.NavigateTo({"INPUT":"","screenRoute":"/form3"})

      Observation: Application navigated to screen 'https://localhost:5001/form3'.

      Thought: Now that I have navigated to the "ModelForm" screen, I should list the currently displayed forms to know what fields to fill out.

      Action: FormCompletionPlugin.ListCurrentlyDisplayedForms({"INPUT":""})

      Observation: This is a list of all of the forms that are currently displayed on the screen: 
      [{"FormName":"Form3Model","FormDescription":"This is a form that lets you register a new user.","FormFields":[{"FieldName":"Username","Description":"The name of the user","FieldType":"String"},{"FieldName":"Email","Description":"The email of the user","FieldType":"String"},{"FieldName":"Password","Description":"The new password for the user. Do not fill this form field.","FieldType":"String"},{"FieldName":"Password2","Description":"The repetition of new password for the user. Do not fill this form field.","FieldType":"String"}]}]

      Thought: I can see the "Form3Model" form with fields for "Username", "Email", "Password", and "Password2". I should fill out the form with the provided information, but I won't fill the "Password" and "Password2" fields as the description says not to.

      Action: No action to take

      Thought: [ACTION] {"action": "FormCompletionPlugin.CompleteForms","action_variables": {"INPUT": "{\"Form3Model\": {\"Username\": \"Anton\", \"Email\": \"anton@test.de\"}}"}

      Action: No action to take

      Thought: 

      Action: No action to take

      Thought: 

      Action: No action to take

      Final Answer: Ich habe die Registrierungsseite geöffnet und das Formular mit dem Benutzernamen Anton und der E-Mail-Adresse anton@test.de ausgefüllt. Bitte geben Sie das Passwort manuell ein, um den Vorgang abzuschließen.

      Result: Ich habe die Registrierungsseite geöffnet und das Formular mit dem Benutzernamen Anton und der E-Mail-Adresse anton@test.de ausgefüllt. Bitte geben Sie das Passwort manuell ein, um den Vorgang abzuschließen.
```

**Expected behavior**

The Stepwise planner recognizes the action command from the model and executes the plugin function.
The model should then observe that the form fields have been filled out and give its final answer.

**Platform**
 - OS: Windows
 - IDE: VS Code
 - Language: C#
 - Source: Microsoft.SemanticKernel.Planning.StepwisePlanner 0.17.230711.7-preview

**Additional context**

This is then only function call in this sample where the model passes an escaped JSON string as the input parameter into the function.
My idea (but I could be totally wrong): Maybe the regex that parses the thought/action misses the action?

**Comment by gingters** [comment]
Additional info:
It's enough to have a simple pair of curly braces within the INPUT variable value to have the regex NOT match anymore.
Example: `[ACTION] {"action": "FormCompletionPlugin.CompleteForms", "action_variables": {"INPUT": "{}" }}\n`

**Comment by nacharya1** [comment]
@gingters thanks for the detailed infomation and details we will follow up and look into this.

**Comment by ShubhamGargCTH** [comment]
I have the exact same issue. For now, I am asking the planner to encode the generated JSON in base 63 until this is fixed.

**Comment by ShubhamGargCTH** [comment]
@nacharya1 do we have any update on when this will be tackled? is there anything we can help you with in any way? 
we are serializing everything to XML and then deserializing, it works but its a pain :)

**Comment by lemillermicrosoft** [comment]
Sorry for delays on this issue. 0.23+ has a new version and attempted to tackle some of these issues. WIll try and verify but confirmation from individuals here would be appreciated, too.

**Comment by ShubhamGargCTH** [comment]
@lemillermicrosoft tested the fix on 0.23+ and seems to be working fine now. Thanks.
